### PR TITLE
Refactor and add new timed utils.

### PR DESCRIPTION
**Pull request by piotrm50**
# Description of change

Move all timed utils to one package.
Implement TimedTaskExecutor and TimedHeap. 
Make other util components the new TimedHeap implementation.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] My code follows the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/timed/taskexecutor.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// Cancel cancels a queued task.
+func (t *TaskExecutor[T]) Cancel(identifier T) (canceled bool) {
+	t.queuedElementsMutex.Lock()
+	defer t.queuedElementsMutex.Unlock()
+
+	queuedElement, queuedElementExists := t.queuedElements.Get(identifier)
+	if !queuedElementExists {
+		return
```
Comment (muXxer) [inline core/timed/taskexecutor.go:81]:
```suggestion
		return false
```
I hate naked returns :trollface:

**Review by muXxer — APPROVED** [review]

File: core/timed/queue.go [hunk 2]
```diff
@@ -278,62 +274,19 @@ const (
...
-// Option //////////////////////////////////////////////////////////////////////////////////////////////////////////////
-
-// Option is the type for functional options of the TimedQueue.
-type Option func(queue *TimedQueue)
+// region Options///////////////////////////////////////////////////////////////////////////////////////////////////////
 
-// WithMaxSize is an Option for the TimedQueue that allows to specify a maxSize of the queue.
-func WithMaxSize(maxSize int) Option {
-	return func(queue *TimedQueue) {
+// WithMaxSize is an ExecutorOption for the TimedQueue that allows to specify a maxSize of the queue.
```
Comment (jonastheis) [inline core/timed/queue.go:287]:
```suggestion
// WithMaxSize is an Option for the timed.Queue that allows to specify a maxSize of the queue.
```

File: core/timed/executor.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/iotaledger/hive.go/core/generics/options"
+)
+
+// region TimedExecutor ////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Executor defines a scheduler that executes tasks in the background at a given time. It does not spawn any
+// additional goroutines for each task and executes the tasks sequentially (in each worker).
+type Executor struct {
+	workerCount  int
+	maxQueueSize int
```
Comment (jonastheis) [inline core/timed/executor.go:16]: maybe naming these optsX is a bit more readable

File: core/timed/taskexecutor.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+// NewTaskExecutor is the constructor of the TaskExecutor.
+func NewTaskExecutor[T comparable](workerCount int) *TaskExecutor[T] {
+	return &TaskExecutor[T]{
+		Executor:       NewExecutor(workerCount),
+		queuedElements: shrinkingmap.New[T, *QueueElement](),
+	}
+}
+
+// ExecuteAfter executes the given function after the given delay.
+func (t *TaskExecutor[T]) ExecuteAfter(identifier T, callback func(), delay time.Duration) *ScheduledTask {
```
Comment (jonastheis) [inline core/timed/taskexecutor.go:29]: Isn't that just exactly the same as `ExecuteAt` with `time.Now().Add(delay)` ? So we can get rid of the duplicate code

**Review by jonastheis — CHANGES_REQUESTED** [review]

**Review by muXxer — APPROVED** [review]

**Review by jonastheis — APPROVED** [review]

**Pull request merged**